File: src/protocol/src/main/java/com/kdyncs/dragonsong/protocol/networking/NetworkWriter.java
/*
 * Copyright (C) 2021 Kingston Dynamics Inc.
 *
 * This file is part of DragonSong
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.kdyncs.dragonsong.protocol.networking;

import com.kdyncs.dragonsong.protocol.utils.Byteinator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.DataOutputStream;
import java.io.IOException;
import java.util.concurrent.LinkedBlockingQueue;

public class NetworkWriter implements Runnable {
    
    // Logging
    private static final Logger LOG = LoggerFactory.getLogger(NetworkWriter.class);
    
    // Threading
    private final Thread thread;
    // Output Queue
    private final LinkedBlockingQueue<byte[]> queue;
    private boolean running;
    // Parent
    private NetworkManager manager;
    
    public NetworkWriter(NetworkManager manager) {
        this.manager = manager;
        this.queue = new LinkedBlockingQueue<>();
        this.thread = new Thread(this);
    }
    
    @Override
    public void run() {
        
        LOG.debug("Thread Starting");
        
        while (running) {
            try {
                
                // Queue Up Writable Data
                byte[] data = queue.take();
                
                // Write Message length and Data
                DataOutputStream output = new DataOutputStream(manager.getSocket().getOutputStream());
                
                //Byteinator.intToBytes(data.length);
                LOG.info("Writing Length {} Bytes", data.length);
                output.write(Byteinator.intToBytes(data.length));
                
                LOG.info("Writing Message {} Bytes", data.length);
                output.write(data);
                output.flush();
                
            } catch (InterruptedException ex) {
                LOG.warn("Writing Interrupted");
            } catch (IOException ex) {
                LOG.warn("Writing Issue", ex);
            }
        }
        
        LOG.debug("Thread Stopped");
    }
    
    public void write(byte[] data) {
        queue.add(data);
    }
    
    public void setManager(NetworkManager manager) {
        this.manager = manager;
    }
    
    public void start() {
        LOG.info("Attempting to start Network Writer");
        
        if (thread.isAlive()) {
            LOG.warn("Network Writer is Already Started");
            return;
        }
        
        running = true;
        thread.start();
    }
    
    public void stop() {
        LOG.info("Attempting to stop Network Writer");
        
        if (!thread.isAlive()) {
            LOG.warn("Network Writer is Already Stopped");
            return;
        }
        
        running = false;
        thread.interrupt();
    }
}

File: sdk/unity/sdk/Assets/KingstonDynamics/DragonSong/Protocol/Networking/NetworkWriter.cs
using System.Collections.Concurrent;
using System.Net.Sockets;
using System.Threading;
using KingstonDynamics.DragonSong.Protocol.Utils;

namespace KingstonDynamics.DragonSong.Protocol.Networking
{
    public class NetworkWriter
    {
        // Manager
        private readonly INetworkManager _manager;
        
        // Threading
        private readonly Thread _thread;
        private bool _running;
        
        // Network Stream
        private TcpClient _client;
        private readonly NetworkStream _stream;
        
        // Write Queue
        private readonly BlockingCollection<byte[]> _queue;
        
        public NetworkWriter(INetworkManager manager)
        {
            // Pull Data
            _manager = manager;
            _client = _manager.Socket;
            _stream = _client.GetStream();
            
            // Create Thread
            _thread = new Thread(Run) {IsBackground = true};

            // Create Queue
            _queue = new BlockingCollection<byte[]>();
            
        }

        public void Start()
        {
            // Can't start Thread Twice'
            if (_thread.IsAlive)
            {
                return;
            }

            _running = true;
            _thread.Start();
        }

        public void Stop()
        {
            if (!_thread.IsAlive)
            {
                return;
            }

            _running = false;
            _thread.Abort(); // TODO: Using this is apparently bed practice
        }
        
        private void Run()
        {
            // TODO: Catch IO Exception

            while (_running)
            {
                var data = _queue.Take();
                var length = Byteinator.IntToBytes(data.Length);

                if (_stream.CanWrite)
                {
                    // Write Length
                    _stream.Write(length, 0, length.Length);
                    _stream.Flush();
                
                    // Write Data
                    _stream.Write(data, 0, data.Length);
                    _stream.Flush();
                }
            }
        }

        public void Write(byte[] data)
        {
            _queue.Add(data);
        }
    }
}